todoapp/TODOapp/app/src/main/java/com/nearsoft/androidschool/todoapp/activities/detail/DetailActivity.java

```diff
@@ -108,22 +123,55 @@ public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                 dateCardView.setVisibility(isChecked ? View.VISIBLE : View.INVISIBLE);
             }
         });
+        locationSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
+            @Override
+            public void onCheckedChanged(CompoundButton compoundButton, boolean isChecked) {
+                if (isChecked) {
+                    if (locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
+                        getLocationFromService();
+                    } else {
+                        showLocationNeededDialog();
+                    }
+                }else{
+                    todoItem.setLat(0.0d);
+                    todoItem.setLng(0.0d);
+                }
+            }
+        });
+    }
+
+    private void getLocationFromService() {
+        if (ActivityCompat.checkSelfPermission(DetailActivity.this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
+            Location location = locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER);
+            saveLocationInToDo(location);
+        } else {
+            ActivityCompat.requestPermissions(DetailActivity.this,
+                    new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, REQUEST_LOCATION_PERMISSION);
+        }
+    }
+
+    private void saveLocationInToDo(Location location) {
+        if (location != null) {
+            todoItem.setLat(location.getLatitude());
+            todoItem.setLng(location.getLongitude());
+        }
     }
 
     private void buttonViewConfig(boolean isEditClicking) {
         titleEditTextView.setEnabled(isEditClicking);
         notesEditTextView.setEnabled(isEditClicking);
         dateCardView.setEnabled(isEditClicking);
         dateSwitch.setEnabled(isEditClicking);
+        locationSwitch.setEnabled(isEditClicking);
         editFab.setVisibility(isEditClicking ? View.INVISIBLE : View.VISIBLE);
         saveFab.setVisibility(isEditClicking ? View.VISIBLE : View.INVISIBLE);
     }
 
-    private void saveData() throws AssertionError {
+    private void saveData() throws Exception {
 //        TODO:this method will make the saving
         String selectedDate = dateTextView.getText().toString();
         if (selectedDate.equals(getText(R.string.date))) {
-            throw new AssertionError(dateCardView.getClass().getSimpleName() + getString(R.string.set_date_or_change_switch_message));
+            throw new Exception(dateCardView.getClass().getSimpleName() + getString(R.string.set_date_or_change_switch_message));
```

Comment:
@jmsalcido the point of the exception was to manage the case when the use activated the switchview but no date was selected, the exception would be thrown, IMHO it is better for this course to simple use a snackbar and return to get out of the method